rig/routing_table/minimise.py

```diff
@@ -0,0 +1,557 @@
+        # aliases dictionary.
+        routing_table, aliases = merge.apply(aliases)
+
+    # Return the finished routing table, it's up to the caller to determine if
+    # the length is appropriate.
+    return routing_table
+
+
+def get_generality(key, mask):
+    """Count the number of Xs in the key-mask pair.
+
+    For example, there are 32 Xs in ``0x00000000/0x00000000``::
+
+        >>> get_generality(0x0, 0x0)
+        32
+
+    And no Xs in ``0xffffffff/0xffffffff``::
+
+        >>> get_generality(0xffffffff, 0xffffffff)
+        0
+    """
+    xs = (~key) & (~mask)
+    return sum(1 for i in range(32) if xs & (1 << i))
+
+
+def get_all_merges(routing_table):
+    """Get possible sets of entries to merge.
+
+    Yield
+    -----
+    :py:class:`~.Merge`
+    """
+    # Memorise entries that have been considered as part of a merge
+    considered_entries = set()
+
+    for i, entry in enumerate(routing_table):
+        # If we've already considered this entry then skip
+        if i in considered_entries:
+            continue
+
+        # Construct a merge by including other routing table entries below this
+        # one which have equivalent routes.
+        merge = set([i])
+        merge.update(
+            j for j, other_entry in enumerate(routing_table[i+1:], start=i+1)
+            if entry.route == other_entry.route
+        )
+
+        # Mark all these entries as considered
+        considered_entries.update(merge)
+
+        # If the merge contains multiple entries then yield it
+        if len(merge) > 1:
+            yield Merge(routing_table, merge)
+
+
+def get_insertion_index(routing_table, generality):
+    """Determine the index in the routing table where a new entry should be
+    inserted.
+    """
+    # Wrapper for get_generality which accepts a routing entry
+    def gg(entry):
+        return get_generality(entry.key, entry.mask)
+
+    # Perform a binary search through the routing table
+    bottom = 0
+    top = len(routing_table)
+    pos = (top - bottom) // 2
+
+    pg = gg(routing_table[pos])
+    while pg != generality and bottom < pos < top:
+        if pg < generality:
+            bottom = pos  # Move up
+        else:  # pg > generality
+            top = pos  # Move down
+
+        # Compute a new position
+        pos = bottom + (top - bottom) // 2
+        pg = gg(routing_table[pos])
+
+    while (pos < len(routing_table) and
+           gg(routing_table[pos]) <= generality):
+        pos += 1
+
+    return pos
+
+
+class Merge(namedtuple("Merge", ["routing_table", "entries", "key", "mask",
+                                 "generality", "goodness",
+                                 "insertion_index"])):
+    """Represents a potential merge of routing table entries.
+    """
+    def __new__(cls, routing_table, entries=set()):
+        # Generate the new key and mask
+        ones = 0x00000000  # Wherever there is a 1 in *any* of the keys
+        key = 0xffffffff  # ... 1 in *all* of the keys
```

Comment:
I've switched these to `any_ones` and `all_ones`, and have changed `mask` to `all_set`.